src/coreclr/src/jit/morph.cpp

```diff
@@ -13705,24 +13715,55 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
+                        bool matchingTypes  = (lclType == typ);
+                        bool matchingWidths = (genTypeSize(lclType) == genTypeSize(typ));
+                        bool matchingFloat  = (varTypeIsFloating(lclType) == varTypeIsFloating(typ));
+
+                        // TYP_BOOL and TYP_UBYTE are also matching types
+                        if (!matchingTypes)
                         {
-                            tree->gtType = typ = temp->TypeGet();
-                            foldAndReturnTemp  = true;
+                            if ((lclType == TYP_BOOL) && (typ == TYP_UBYTE))
+                            {
+                                matchingTypes = true;
+                            }
+                            else if ((lclType == TYP_UBYTE) && (typ == TYP_BOOL))
+                            {
+                                matchingTypes = true;
+                            }
                         }
-                        else if (!varTypeIsStruct(typ) && (lvaTable[lclNum].lvType == typ) &&
-                                 !lvaTable[lclNum].lvNormalizeOnLoad())
+
+                        // If the sizes match and we dont have a float or TYP_STRUCT,
+                        // then we will fold the IND/ADDR and use the LCL_VAR directly
+                        //
+                        if (matchingWidths && matchingFloat && (typ != TYP_STRUCT))
                         {
+                            // For some small types we might need to change to normalize loads or insert a cast here
+                            //
+                            if (varTypeIsSmall(typ))
+                            {
+                                // For any stores of small types, we will force loads to be normalized
+                                // this is necessary as we need to zero/sign extend any load
+                                // after this kind of store.
+                                //
+                                if (isStore)
+                                {
+                                    varDsc->lvForceLoadNormalize = true;
+                                }
+                                // otherwise we have a load operation
+                                //
+                                // Do we have a non matching small type load?
+                                // if so we need to insert a cast operation
+                                else if (!matchingTypes)
+                                {
+                                    addCastToType = typ;
+                                }
+                            }
+                            // we will fold the IND/ADDR and reduce to just the local variable
+                            //
                             tree->gtType = typ = temp->TypeGet();
                             foldAndReturnTemp  = true;
                         }
-                        else
+                        if (!foldAndReturnTemp)
```

Comment:
Does `else`->`if (!foldAndReturnTemp)` change something that I am missing?

Reply:
While working on this with some of my previous changes the block above didn't always set `foldAndReturnTemp `to true.